core/modules/loader/BufferUdp.cc

```diff
@@ -0,0 +1,158 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2018 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+// Class header
+#include "BufferUdp.h"
+
+// system headers
+
+
+// Third-party headers
+
+
+// qserv headers
+#include "loader/LoaderMsg.h"
+
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.loader.BufferUdp");
+}
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+/// Repeatedly read a socket until a valid MsgElement is read, eof,  or an error occurs.
+MsgElement::Ptr BufferUdp::readFromSocket(boost::asio::ip::tcp::socket& socket, std::string const& note) {
+    for (;;) {
+        LOGS(_log, LOG_LVL_DEBUG, note << " readFromSocket");
+        boost::system::error_code error;
+
+        // If there's something in the buffer already, get it and return.
+        // This can happen when the previous read of socket read multiple elements.
+        MsgElement::Ptr msgElem = _safeRetrieve();
+        if (msgElem != nullptr) {
+            return msgElem;
+        }
+
+        size_t len = socket.read_some(boost::asio::buffer(_wCursor, getAvailableWriteLength()), error);
+        _wCursor += len; /// must advance the cursor.
```

Comment:
This may not be safe to do before the status code is analyzed. The method may return garbage in case of errors.

Reply:
I need to move the cursor in the buffer so the data can be parsed. The cursor only indicates the extent of the data written into the buffer by read_some. If there's an error code, this function will throw before trying to parse the data.